src/sbuf/SBuf.cc

```diff
@@ -148,7 +148,7 @@ SBuf::rawAppendFinish(const char *start, size_type actualSize)
     debugs(24, 8, id << " finish appending " << actualSize << " bytes");
 
     size_type newSize = length() + actualSize;
-    Must2(newSize <= min(maxSize,store_->capacity-off_), "raw append overflow");
+    Must3(newSize <= min(maxSize, store_->capacity-off_), "raw append fits", Here());
     len_ = newSize;
     store_->size = off_ + newSize;
```

Comment:
FTR, this code needs to be completely rewritten IMO, and that rewrite is likely to remove all the Musts, but I am limiting this PR to fixing/replacing Must2() API, not fixing any existing Must2() users.

Reply:
Can't we just drop this rarely used alternative to Must() ?

Reply:
We could, but that feels like a step backwards to me. The reason we and C++ has the second argument is to convert these typical use cases:

```C++
    Must(a < b);
```

... which forces the reader to figure out what "a < b" really means and forces the developer to figure out which of the three "a < b" assertions has failed in some unknown code (where line numbers do not help much);


```C++
    Must(a < b); // the caller gave as a malformed c value
```
... which solves the first problem but still forces the developer to figure out which of the three "a < b" assertions has failed in some unknown code (where line numbers do not help much);


```C++
    Must((a < b) && "the caller gave as a malformed c value");
```

... which solves the above problems but at the expense of a more complex expression that confuses newbies (Squid does not use this approach, but many projects do);

to 

```C++
    Must(a < b, "the caller gave as a malformed c value");
```
which solves all the problems above. I bet that is exactly why C++11 added the message argument to static_assertion (they could not do it to the C assert() macro for obvious reasons).

My only doubt here is that the current post-C++20 proposal does not add the message argument to C++ asserts: `[[ a < b ]]`. Perhaps they will do that eventually or perhaps they just had to sacrifice the message in order to use the `[[ ... ]]` syntax. If you can find why they did not add the message, please let me know!

Just to make sure we are on the same page regarding the proposed "drop" option consequences: If we drop Must-with-a-message, would you be OK to see `Must(c && "explanation")` replacements?

Reply:
> 
> My only doubt here is that the current post-C++20 proposal does not add the message argument to C++ asserts: `[[ a < b ]]`. Perhaps they will do that eventually or perhaps they just had to sacrifice the message in order to use the `[[ ... ]]` syntax. If you can find why they did not add the message, please let me know!
> 


> Just to make sure we are on the same page regarding the proposed "drop" option consequences: If we drop Must-with-a-message, would you be OK to see `Must(c && "explanation")` replacements?

IIRC there were compiler issues with the second argument never being able to be false. Static analysis highlights that as a bug.  I think we should rename the variables used to something that helps self-document when only condition is visible. sed /newSize/newSbufSize/ or such.

Reply:
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3928.pdf looks like the C++-17 document making static_assert string becoming optional. Unfortunately it just lists a lot of requested changes with little or no rational behind them.

The C++23 document ("current post-C++-20 proposal" ?)  is scoped as fixing only the issue of compile errors when comma is used in syntax of conditions passed as assert() parameter. It is seeking to be explicitly backward compatible with current syntax usage, and the C language definition of assert() as a macro.

Reply:
> http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3928.pdf looks like the C++-17 document making static_assert string becoming optional. Unfortunately it just lists a lot of requested changes with little or no rational behind them.

I bet the second static_assert argument was made optional to avoid code duplication. I have seen text alluding to that in Sutter slides IIRC. After the two-argument static_assert was added, folks noticed a lot of new code that looked like this:

```C++
static_assert(c, "c");
```

Making the second argument optional was the right solution. To avoid misunderstanding, making the second argument optional does not imply the second argument should not have been added in the first place. There are two different problems here, and, as of C++17, both are solved.

> The C++23 document ("current post-C++-20 proposal" ?) is scoped as fixing only the issue of compile errors when comma is used in syntax of conditions passed as assert() parameter. It is seeking to be explicitly backward compatible with current syntax usage, and the C language definition of assert() as a macro.

The proposed `[[assert]]` language feature I was talking about is mentioned in GotW [97](https://herbsutter.com/2021/01/11/gotw-97-solution-assertions-difficulty-4-10/). It is documented in a contract-based programming support proposal [P0542R5](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0542r5.html). While there is support for contract_violation::comment(), that comment looks compiler-generated rather than developer-specified:

```C++
// Returns: Implementation-defined text describing the predicate of the violated contract.
string_view comment() const noexcept;
```

I do not know what the current `[[assert]]` adoption status is.

Reply:
> > Alex: Just to make sure we are on the same page regarding the proposed "drop" option consequences: If we drop Must-with-a-message, would you be OK to see `Must(c && "explanation")` replacements?

> IIRC there were compiler issues with the second argument never being able to be false. 

You lost me here. If we drop Must-with-a-message, there will be no second argument.

> Static analysis highlights that as a bug.

If you are thinking about `Must(false)` and equivalent, then that is a different/unrelated topic.

If you are thinking that `Must(c && "c")` will cause problems with static analysis, then that is one more reason to keep support for the message argument in Must().

> I think we should rename the variables used to something that helps self-document when only condition is visible. sed /newSize/newSbufSize/ or such.

While self-documenting code is the best outcome we can hope for, most developers will have a negative reaction to refactoring requests motivated by Must() message clarity concerns. Moreover, such refactoring is not always possible or desirable. There are legitimate reasons for using source code comments, of course, and the second Must() argument (or equivalent) is very similar to a source code comment.

Reply:
> > > Alex: Just to make sure we are on the same page regarding the proposed "drop" option consequences: If we drop Must-with-a-message, would you be OK to see `Must(c && "explanation")` replacements?
> 
> > IIRC there were compiler issues with the second argument never being able to be false.
> 
> You lost me here. If we drop Must-with-a-message, there will be no second argument.
> 

Sorry, I meant the second boolean argument. The quoted-string part.

We used to have a few (A && "string") conditions in the code. Static analysis tools complained that the "string" part of the condition could never be false and output a bunch of false positives. Apart from that annoyance I am okay with the construct.

> > Static analysis highlights that as a bug.
> 
> If you are thinking about `Must(false)` and equivalent, then that is a different/unrelated topic.
> 
> If you are thinking that `Must(c && "c")` will cause problems with static analysis, then that is one more reason to keep support for the message argument in Must().
> 
> > I think we should rename the variables used to something that helps self-document when only condition is visible. sed /newSize/newSbufSize/ or such.
> 
> While self-documenting code is the best outcome we can hope for, most developers will have a negative reaction to refactoring requests motivated by Must() message clarity concerns. Moreover, such refactoring is not always possible or desirable. There are legitimate reasons for using source code comments, of course, and the second Must() argument (or equivalent) is very similar to a source code comment.

Good points.  I think I am convinced now that we do need a separate Must-with-message - as much of a hack as it is.

Please do seek to prefer new code that does not need it though. Must() should remain our preferred name, one day we may be able to overload it and remove the macro naming hack.

Reply:
> Please do seek to prefer new code that does not need it though.

Yes, with all other factors being equal, less code is better (and self-documenting code is better than the code relying on explicit messages or comments for documentation).